src/cdk/tree/tree.ts

```diff
@@ -1172,6 +1190,13 @@ export class CdkTreeNode<T, K = T> implements OnDestroy, OnInit, TreeKeyManagerI
   readonly _dataChanges = new Subject<void>();
 
   private _inputIsExpandable: boolean = false;
+  /**
+   * Flag used to determine whether or not we should be focusing the actual element based on
+   * some user interaction (click or focus). On click, we don't forcibly focus the element
+   * since the click could trigger some other component that wants to grab its own focus
+   * (e.g. menu, dialog).
+   */
+  private _shouldFocus = true;
```

Comment:
I don't think I understand why we wouldn't want to focus on click. Native elements usual focus themselves when clicked.

Do we have an example of other component that grabs its own focus?

Reply:
native elements have focus by virtue of the browser automatically leaving focus on the last interacted element; they do not _force_ focus to be on them. the problem here is essentially:
```
<mat-tree-node
  [matMenuTriggerFor]="menu">
</mat-tree-node>
```

in this scenario, without this change:
1. mat-menu tries to focus the first menu item when it opens
2. the tree then steals focus back to the tree item

similar behaviour occurs with dialogs where we take focus back from the dialog.

with this change, if there is no additional behaviour where focus is taken from the tree item, the browser will still leave the tree item as the currently active element since it's marked as focusable via `tabindex="0"`